### DISCUSSION TITLE
How to log and edit data that comes through WebSockets

### DISCUSSION THREAD

**Discussion by Mistera91** [synchronizing/mitm#27] [Q&A] [answered]
Hi,
The title is pretty self-explanatory, I can't find a way to get the packets in real time and edit them (e.g. apply a filter to replace "foo" by "test" in the payload, or just log the payload to some file).
Maybe it is the purpose of mitm.core.Middleware class ? Or is it even possible ?

**Comment by synchronizing** [comment, marked as answer]
How are you attempting to go about doing this? The `Middleware` extension can indeed be helpful for this.

**Reply by Mistera91** [reply]
I think that my problem is that the https traffic is not going through the proxy (by the way, I use ProxyDroid). I am currently trying to get the certificate trusted by the system and apps on my phone but ~it is hard~ I suck at it.

**Reply by Mistera91** [reply]
Ok, so https works but I can only see the WebSocket initializing and closing. Is there a way to also get the bidirectional message ?

**Reply by synchronizing** [reply]
It should indeed be possible to do what you are trying to do. How are you using `mitm`? If you simply want to intercept and view traffic I would recommend checking out the [`mitmproxy`](https://mitmproxy.org/) project. `mitm` was/is intended for projects that require interception of TCP messages, and less so analytics.

**Reply by Mistera91** [reply]
I use `mitm` inside a custom instance of the `Middleware` class, then I log the data using `print(data)` in each `async` function. This logs all the traffic (except for websockets bidirectional messages) and also allows to edit it with the `return data` at the end. I think I will try to use `mitmproxy`, I've tried it once but it didn't worked and I kinda left it behind.
Thanks for your project, I've already managed to create cool things with it (like one that replaces the data of an app so that I can fake proof in front of someone that the app is wrong), and these were very fun to create !

**Reply by synchronizing** [reply]
Forgive me for the late reply - holidays and all. 

Awesome, happy to hear you've been using `mitm`! I will have to run some tests with Websockets to see how its functioning with it. In theory, it should work out of the box, but unsure. Worst case creating a `Protocol` class that is capable of intercepting and interpreting HTTP(S) + Websockets might be the way.

**Reply by Mistera91** [reply]
Oh ok cool ! I'm trying to switch to `mitmproxy` but I again struggled with openssl issues for decades so I kinda let it away. Happy to hear that you will try this out, your python module makes it so much easier !

By the way, if you have any idea on how to make it work with mobile data on android (I have root access if needed), I would really appreciate it !

**Reply by synchronizing** [reply]
> ... your python module makes it so much easier !

Glad to hear you feel this way. The reason I built this project was because I was also trying to built on top of `mitmproxy` and felt the project's interfaces were overly complicated, so I know where you are coming from.

> By the way, if you have any idea on how to make it work with mobile data on android (I have root access if needed), I would really appreciate it !

Specify `mitm` CA folder to dump the private key & certificate (see docs [here](https://synchronizing.github.io/mitm/docs/trusting-mitm.html)). Transfer over only the certificate to your Android phone, and install it on the system. Then - making sure you are on the same wifi as your computer - go to Network settings and set your proxy to your computers LAN IP address. In some cases, like an iPhone, you must give the certificate full authority.

**Reply by Mistera91** [reply]
> Then - making sure you are on the same wifi as your computer

That's what I am doing so far to connect to the proxy, but I want to be able to connect to it from wherever I want (through mobile data).
I have 3 main ideas on how to do this :
- Execute the python script on the mobile device and connect to localhost (but how to connect to localhost ?) (I found that PCAPdroid uses `mitmproxy` and can connect to the app's builtin script locally, and that it works on mobile data)

- Execute it on a server, and connect the mobile device to the server (but how to connect to sevrer ?)

- Execute it on a server but with a VPN (like `mitmproxy`'s wireguard mode), because connecting to a VPN is much easier (builtin android) (but I guess it would require a lot to create this)

**Reply by synchronizing** [reply]
I've never personally experimented with connecting to `mitm` via cellular network, so not sure if I can be of help here. `mitm` doesn't provide VPN support out-of-the-box.
